timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimeLockServer.java

```diff
@@ -176,49 +182,80 @@ private LeaderConfig getLeaderConfig(TimeLockServerConfiguration configuration)
-                client);
-
-        Supplier<AsyncTimelockService> rawAsyncTimelockServiceSupplier = () -> {
-            ScheduledExecutorService reaperExecutor = Executors.newSingleThreadScheduledExecutor(
-                    new ThreadFactoryBuilder()
-                            .setNameFormat("async-lock-reaper-" + client + "-%d")
-                            .setDaemon(true)
-                            .build());
-            ScheduledExecutorService timeoutExecutor = Executors.newSingleThreadScheduledExecutor(
-                    new ThreadFactoryBuilder()
-                            .setNameFormat("async-lock-timeouts-" + client + "-%d")
-                            .setDaemon(true)
-                            .build());
-            return new AsyncTimelockServiceImpl(
-                    AsyncLockService.createDefault(reaperExecutor, timeoutExecutor),
-                    rawTimestampServiceSupplier.get());
-        };
-
+    private AsyncTimelockService createAsyncTimelockService(
+            String client,
+            Supplier<ManagedTimestampService> rawTimestampServiceSupplier) {
         return AwaitingLeadershipProxy.newProxyInstance(
                 AsyncTimelockService.class,
-                rawAsyncTimelockServiceSupplier,
+                () -> createRawAsyncTimelockService(client, rawTimestampServiceSupplier),
                 leaderElectionService);
     }
 
+    private AsyncTimelockService createRawAsyncTimelockService(
+            String client,
+            Supplier<ManagedTimestampService> timestampServiceSupplier) {
+        ScheduledExecutorService reaperExecutor = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("async-lock-reaper-" + client + "-%d")
+                        .setDaemon(true)
+                        .build());
+        ScheduledExecutorService timeoutExecutor = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("async-lock-timeouts-" + client + "-%d")
+                        .setDaemon(true)
+                        .build());
+        return new AsyncTimelockServiceImpl(
+                AsyncLockService.createDefault(reaperExecutor, timeoutExecutor),
+                timestampServiceSupplier.get());
+    }
+
+    private TimeLockServices createLegacyTimeLockServices(
+            Supplier<ManagedTimestampService> rawTimestampServiceSupplier,
+            RemoteLockService lockService) {
+        TimelockService timelockService = new LockTokenConvertingTimelockService(
+                new LegacyTimelockService(rawTimestampServiceSupplier.get(),
+                        lockService,
+                        LockClient.of("legacy")));
+
+        return TimeLockServices.create(
+                new TimelockTimestampServiceAdapter(timelockService),
```

Comment:
Actually this is better because it allows me to get rid of the additional proxy I had for the managed service as well